### PR TITLE
Avoid useless context copy when building query interpreters

### PR DESCRIPTION
**Pull request by amosbird**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Avoid useless context copy when building query interpreters.

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

File: src/Interpreters/IInterpreterUnionOrSelectQuery.h [hunk 1]
```diff
@@ -12,8 +12,13 @@ class IInterpreterUnionOrSelectQuery : public IInterpreter
...
 public:
     IInterpreterUnionOrSelectQuery(const ASTPtr & query_ptr_, ContextPtr context_, const SelectQueryOptions & options_)
+        : IInterpreterUnionOrSelectQuery(query_ptr_, Context::createCopy(context_), options_)
+    {
+    }
+
+    IInterpreterUnionOrSelectQuery(const ASTPtr & query_ptr_, ContextMutablePtr context_, const SelectQueryOptions & options_)
         : query_ptr(query_ptr_)
-        , context(Context::createCopy(context_))
+        , context(context_)
```
Comment (amosbird) [inline src/Interpreters/IInterpreterUnionOrSelectQuery.h:21]: Sometimes we already have mutable context (e.g. `InterpreterFactory::get`). There is no need to create a copy.

File: src/Interpreters/InterpreterSelectWithUnionQuery.cpp [hunk 2]
```diff
@@ -31,8 +31,13 @@ namespace ErrorCodes
 }
 
 InterpreterSelectWithUnionQuery::InterpreterSelectWithUnionQuery(
-        const ASTPtr & query_ptr_, ContextPtr context_,
-        const SelectQueryOptions & options_, const Names & required_result_column_names)
+    const ASTPtr & query_ptr_, ContextPtr context_, const SelectQueryOptions & options_, const Names & required_result_column_names)
+    : InterpreterSelectWithUnionQuery(query_ptr_, Context::createCopy(context_), options_, required_result_column_names)
```
Comment (rschu1ze) [inline src/Interpreters/InterpreterSelectWithUnionQuery.cpp:35]: Perhaps it is more consistent to provide another ctor to IInterpreterUnionOrSelectQuery which does the copying?

File: src/Interpreters/IInterpreterUnionOrSelectQuery.h (hunk 1 above)
Comment (amosbird) [reply src/Interpreters/IInterpreterUnionOrSelectQuery.h:21]: >  it looks to me like the original context is supposed to stay unchanged.

It should be fine. There are lots of interpreters take mutable reference of context, e.g. InterpreterDropQuery.

File: src/Interpreters/InterpreterSelectWithUnionQuery.cpp (hunk 2 above)
Comment (amosbird) [reply src/Interpreters/InterpreterSelectWithUnionQuery.cpp:35]: > provide another ctor to IInterpreterUnionOrSelectQuery which does the copying

It's already done. And yes, we can use that c'tor to be more consistent.

Comment (amosbird) [reply src/Interpreters/InterpreterSelectWithUnionQuery.cpp:35]: Let's see if templating all related c'tors make it better)

Comment (amosbird) [reply src/Interpreters/InterpreterSelectWithUnionQuery.cpp:35]: I guess no. We're allowed to implicitly convert `ContextMutablePtr` to `ContextPtr`. Template will introduce ambiguity.

**Review by rschu1ze — APPROVED** [review]

**Comment by amosbird** [comment]
> Do you have the rights to merge? 

I don't.

> If not, I can also hit the magic button.

Thank you.

**Pull request merged**